src/Poseidon/Chronicle.hs

```diff
@@ -0,0 +1,184 @@
+
+import           Poseidon.Package        (PoseidonPackage (..))
+import           Poseidon.SecondaryTypes (HasNameAndVersion (..),
+                                          VersionComponent (..),
+                                          updateThreeComponentVersion)
+import           Poseidon.Utils          (PoseidonException (..), PoseidonIO)
+
+import           Control.Monad.Catch     (throwM)
+import           Control.Monad.IO.Class  (liftIO)
+import           Data.Aeson              (FromJSON, ToJSON, object, parseJSON,
+                                          toJSON, withObject, (.!=), (.:),
+                                          (.:?), (.=))
+import qualified Data.ByteString         as B
+import           Data.Function           (on)
+import           Data.List               (elemIndex)
+import           Data.Maybe              (fromMaybe)
+import qualified Data.Set                as S
+import           Data.Time               (Day, UTCTime (..), getCurrentTime)
+import           Data.Version            (Version, makeVersion)
+import           Data.Yaml               (decodeEither')
+import           Data.Yaml.Pretty        (defConfig, encodePretty,
+                                          setConfCompare, setConfDropNull)
+import           GitHash                 (getGitInfo, giHash)
+import           System.Directory        (createDirectoryIfMissing,
+                                          makeAbsolute)
+import           System.FilePath         (makeRelative, takeDirectory)
+
+data PoseidonPackageChronicle = PoseidonPackageChronicle
+    { snapYamlTitle            :: String
+    , snapYamlDescription      :: Maybe String
+    , snapYamlChronicleVersion :: Version
+    , snapYamlLastModified     :: Day
+    , snapYamlPackages         :: S.Set PackageIteration
+    }
+    deriving (Show, Eq)
+
+instance FromJSON PoseidonPackageChronicle where
+    parseJSON = withObject "PoseidonYamlStruct" $ \v -> PoseidonPackageChronicle
+        <$> v .:  "title"
+        <*> v .:? "description"
+        <*> v .:  "chronicleVersion"
+        <*> v .:  "lastModified"
+        <*> v .:? "packages" .!= S.empty
+
+instance ToJSON PoseidonPackageChronicle where
+    toJSON x = object $ [
+        "title"            .= snapYamlTitle x,
+        "description"      .= snapYamlDescription x,
+        "chronicleVersion" .= snapYamlChronicleVersion x,
+        "lastModified"     .= snapYamlLastModified x] ++
+        if not $ null (snapYamlPackages x) then ["packages" .= snapYamlPackages x] else []
+
+-- | A data type to represent a package state
+data PackageIteration = PackageIteration
+    { pacStateTitle   :: String   -- ^ the title of the package
+    , pacStateVersion :: Version  -- ^ the version of the package
+    , pacStateCommit  :: String   -- ^ the hash of a relevant commit where a package can be accessed in this version
+    , pacStatePath    :: FilePath -- ^ the file path where the POSEIDON.yml file is stored
```

Comment:
Do we actually need this `pacStatePath`? It seems to lead to this ugly thing that you have to carry the path to the chronicle file into all functions, like `makeChronicle` and `chroniclePackages`, where it seems counter-intuitive why it is needed. Especially since you still require an output file when you actually write the chronicle (which makes sense). I would love to reduce that kind of redundancy.

Reply:
OK, I see that you use this for the git checkout. But I think it's actually not necessary. In principle, a package should be findable and uniquely determined by i) a git commit, and ii) a name and version. I think storing the path with the snapshot is kind of brittle...

Reply:
Hm. But then we have to rely on the package title to be identical to the directory name, right? Or we crawl to POSEIDON.yml files all the time. But that probably involves more code then just encoding the paths... I'll think about it.

Reply:
I think crawling the directories for the right packages is the right thing to do, and definitely more robust than storing a relative path. I don't think it will be more code. Just one function to pull out the right package.